### PR TITLE
Fix resizing worlds crashing the world editor

### PR DESCRIPTION
**Pull request by AzureBerylBlue**
The resize world code was doing lots of strange things, like deleting some vehicles/warps, then deleting them all later when New was called, so I just rewrote it. The code that was crashing is the code I deleted from worldeditor.cpp.

### CODE REVIEW

**Comment by mmatyas** [comment]
Hm I tried resizing worlds but it seems to work fine for me? How exactly this crash happens? Tried putting down some vehicles, they do seemed to disappear when the resize dialog opened but then they reappeared again after the change (looks like they are stored twice?). Trying out your changes, I've also noticed a bug where vehicles outside the new world borders don't get removed:

Screenshot_2019-11-09_11-26-44x

**Comment by AzureBerylBlue** [comment]
You're right, the vehicles are stored in both a vector and an array, and so are the warps. The code that was crashing was those iterators in worldeditor.cpp. Removing them fixes the crash. The Resize() function had to be rewritten because the old version called New(), which deleted all the stages. (If you pressed E, everything was gone.)

**Comment by AzureBerylBlue** [comment]
Ok, I fixed it. The problem was RemoveVehicleFromTile() was modifying the list that was being iterated over.

**Comment by mmatyas** [comment]
Ok; now that we know what causes the crash, is there any part of the patch that's no longer necessary (eg. rewriting the warp list freeing)? If possible I'd prefer avoiding code duplication with `New`, do you have some good idea for that?

File: src/smw/world.cpp [hunk 1]
```diff
@@ -1065,46 +1065,35 @@ void WorldMap::New(short w, short h)
...
-                tiles[iCol][iRow].iConnectionType = tempTiles[iCol][iRow].iConnectionType;
-                tiles[iCol][iRow].iType = tempTiles[iCol][iRow].iType;
-            }
-        }
-
-        for (short iCol = 0; iCol < w; iCol++)
-            delete [] tempTiles[iCol];
-
-        delete [] tempTiles;
-    }
```
Comment (mmatyas) [inline src/smw/world.cpp]: I suppose the temp tiles should be still freed here.

File: src/worldeditor/worldeditor.cpp [hunk 2]
```diff
@@ -4779,21 +4779,17 @@ int resize_world()
...
-		std::vector<WorldWarp*>::iterator itrWarp = warplist.begin(), limWarp = warplist.end();
-        while (itrWarp != limWarp) {
-			delete (*itrWarp);
-
-			itrWarp = warplist.erase(itrWarp);
-			limWarp = warplist.end();
-
-			itrWarp++;
-		}
+		warplist.clear();
```
Comment (mmatyas) [inline src/worldeditor/worldeditor.cpp]: This only deletes the pointers, not the pointed data. Is that intentional?

**Comment by AzureBerylBlue** [comment]
Fixed

**Comment by mmatyas** [comment]
Ok seems to be working now, would you like to work some more on reducing the code duplication between Resize and New (`//Copy tiles to new map` part) or it's ready for merging?

**Comment by AzureBerylBlue** [comment]
It's ready for merging. The reason New() shouldn't be called is because New() calls Cleanup(), which calls ResetTourStops() and deletes all the stages.

**Pull request merged**